butterfree/pipelines/feature_set_pipeline_metadata.py

```diff
@@ -0,0 +1,220 @@
+            Field(discriminator="type"),
+        ]
+    ] = Field(
+        ...,
+        description="A list of data sources required to generate the feature set",
+    )
+
+    writers: List[WriterMetadata] = Field(
+        ..., description="The writers to be used for the feature set"
+    )
+    catalog: Catalog = Field(..., description="Metadata about the feature set's output")
+
+    @field_serializer("feature_set_pipeline")
+    def serialize_feature_set_pipeline(self, pipeline: FeatureSetPipeline) -> str:
+        """Serialize feature_set_pipeline to its class name."""
+        return pipeline.__class__.__name__
+
+    @classmethod
+    def _get_windows_definition(
+        cls, feature_set_pipeline: FeatureSetPipeline
+    ) -> Optional[List[str]]:
+        """Get windows definition if the feature set is an AggregatedFeatureSet.
+
+        Args:
+            feature_set_pipeline: The FeatureSetPipeline to get windows from.
+
+        Returns:
+            Optional list of window definitions or None if not applicable.
+        """
+        if (
+            isinstance(feature_set_pipeline.feature_set, AggregatedFeatureSet)
+            and feature_set_pipeline.feature_set._windows
+        ):
+            return [
+                window.frame_boundaries.window_definition
+                for window in feature_set_pipeline.feature_set._windows
+            ]
+        return None
+
+    @classmethod
+    def _create_catalog(cls, feature_set_pipeline: FeatureSetPipeline) -> Catalog:
+        """Create a Catalog object from the feature set pipeline.
+
+        Args:
+            feature_set_pipeline: The FeatureSetPipeline to create catalog from.
+
+        Returns:
+            A Catalog instance with feature set metadata.
+        """
+
+        # This method already deals with the colums name transformation
+        catalog_schema = feature_set_pipeline.feature_set.get_schema()
+
+        columns = [
+            Column(
+                name=column["column_name"],
+                data_type=column[
+                    "type"
+                ].__class__.__name__,  # The get_schema method returns pyspark.sql.types
+                description=column["description"],
+                primary_key=column["primary_key"],
+            )
+            for column in catalog_schema
+        ]
+
+        return Catalog(
+            feature_set_name=feature_set_pipeline.feature_set.name,
+            description=feature_set_pipeline.feature_set.description,
+            columns=columns,
+        )
+
+    @classmethod
+    def _is_incremental(cls, feature_set_pipeline: FeatureSetPipeline) -> bool:
+        has_incremental_reader = any(
+            [
+                reader.incremental_strategy
+                for reader in feature_set_pipeline.source.readers
+            ]
+        )
+
+        return has_incremental_reader
+
+    @classmethod
+    def _get_key_features(
```

Comment:
Is "key" here referencing the identifiers for each feature set (like primary keys or composite keys)?

Reply:
Yes. Butterfree calls them that way. But this is a function that I forgot to delete